Index: hdparm-9.43/hdparm.c
===================================================================
--- hdparm-9.43.orig/hdparm.c
+++ hdparm-9.43/hdparm.c
@@ -1503,6 +1503,7 @@ static void usage_help (int clue, int rc
 	" --offset          use with -t, to begin timings at given offset (in GiB) from start of drive\n"
 	" --prefer-ata12    Use 12-byte (instead of 16-byte) SAT commands when possible\n"
 	" --read-sector     Read and dump (in hex) a sector directly from the media\n"
+	" --repair-sector   Alias for the --write-sector option (VERY DANGEROUS)\n"
 	" --security-help   Display help for ATA security commands\n"
 	" --trim-sector-ranges        Tell SSD firmware to discard unneeded data sectors: lba:count ..\n"
 	" --trim-sector-ranges-stdin  Same as above, but reads lba:count pairs from stdin\n"
